### PR TITLE
 Use HMM to seperate node from v8 in cpu 

### PR DESCRIPTION
**Pull request by AndreasMadsen**
When v8 runs GC, Optimization, or something else the CPU graph spikes.
However this has very little to do with the CPU requirements of the
user's application.

As such, if the user has an I/O issue and this causes a lot of
GCScavenger runs from V8, the CPU can spike a lot which hides the I/O
issue. This uses a Hidden Markov Model to statistically separate those
two modes.

PS: I reverted #134 since it didn't really make sense to me (at least it doesn't do what it says on the tin) and the CPU analysis handes noise completely different now. So I think you should reevaluate if it still makes sense.

Also: `@tensorflow/tfjs-node` is an optional native dependency. I'm not sure what your policy is on native dependencies. – If it is not included you get an message about that it is recommended and I can't find an option to turn that off. But properly there is a way, if not we can make a pull request.

/cc @mcollina @jasnell @mafintosh @BridgeAR

### CODE REVIEW

File: analysis/analyse-delay.js [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 const summary = require('summary')
 
-function analyseDelay (processStatSubset, traceEventSubset) {
```
[the remark this answers is not in the document]
Comment (AndreasMadsen) [reply analysis/analyse-delay.js]: It was just for consistency. I can remove it if you prefer.

File: test/analysis-memory.test.js [hunk 2]
```diff
@@ -36,24 +36,24 @@ test('analyse memory - delay correlation', function (t) {
...
-  t.strictDeepEquals(analyseMemory(badMemoryStat, badMemoryGc), {
+  t.strictDeepEquals(await analyseMemory(badMemoryStat, badMemoryGc), {
     external: false,
     heapTotal: false,
     heapUsed: true,
     rss: false
   })
 
   t.end()
-})
```
[the remark this answers is not in the document]
Comment (AndreasMadsen) [reply test/analysis-memory.test.js]: It is what the `node-tap` documentation recommends. https://www.node-tap.org/promises/

File: analysis/index.js [hunk 3]
```diff
@@ -59,20 +59,26 @@ function analysis (traceEventReader, processStatReader) {
...
-      'issueCategory': category
+      result.push({
+        'interval': intervalTime,
+        'issues': issues,
+        'issueCategory': category
+      })
+      result.push(null)
+    })
+    p.catch(function (err) {
+      result.emit('error', err)
```
[the remark this answers is not in the document]
Comment (AndreasMadsen) [reply analysis/index.js]: If we make the parent function `async` then that `Promise` will swallow any thrown error that is not in the try/catch. I think any integration between `async`/`await`/`Promise` and a `callback`/`eventemitter` needs to eventually use a `.catch()` and `.then()`. `hidden-markov-model-tf` uses `async`/`await` because that is what TensorFlow.js uses.

File: analysis/analyse-delay.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcollina) [reply analysis/analyse-delay.js]: +1 for keeping them sync.

File: analysis/index.js [hunk 4]
```diff
@@ -59,20 +59,26 @@ function analysis (traceEventReader, processStatReader) {
...
     }
 
-    const issues = {
-      'delay': analyseDelay(processStatSubset, traceEventSubset),
-      'cpu': analyseCPU(processStatSubset, traceEventSubset),
-      'memory': analyseMemory(processStatSubset, traceEventSubset),
-      'handles': analyseHandles(processStatSubset, traceEventSubset)
-    }
-    const category = issueCategory(issues, {detectNoise: true})
+    const p = Promise.all([
```
Comment (mcollina) [inline analysis/index.js]: I would `await`  here. Can you also refactor this function? Using both `async.parallel` and `Promise.all` seems a bit too hard to read.

Also, the function is too long.

File: analysis/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcollina) [reply analysis/index.js]: Should we use [`util.callbackify`](https://nodejs.org/api/util.html#util_util_callbackify_original) instead?

File: test/analysis-cpu.test.js [hunk 5]
```diff
@@ -1,21 +1,37 @@
...
     const badCPU = generateProcessStat({
       cpu: [50, 40, 10, 10, 100, 50, 40, 10, 10, 30, 10]
     }, noise)
-    t.strictEqual(analyseCPU(badCPU, []), true)
+    t.strictEqual(await analyseCPU(badCPU, []), true)
   }
 
   t.end()
-})
+}).catch(tap.throw)
```
Comment (mcollina) [inline test/analysis-cpu.test.js]: I've used tap extensively without this catch, and I don't think it's needed.

**Comment by mcollina** [comment]
The original https://github.com/nearform/node-clinic-doctor/pull/134 was added because of Zoom. Basically if the CPU is busy doing something else (a Zoom call), the detection was completely wrong.
@mafintosh do you still have around the original profile?

---

A test is failing:

```
    # Subtest: Collect - process stat - cpu usage
        ok 1 - sleep has high usage, usage was: 1.0522941394617236
        not ok 2 - timeout has low usage, usage was: 0.47604811277444914
          ---
          at:
            line: 70
            column: 7
            file: test/collect-process-stat.test.js
            type: Timeout
            method: _onTimeout
          stack: |
            Timeout._onTimeout (test/collect-process-stat.test.js:70:7)
          source: |
            t.ok(sample.cpu >= 0 && sample.cpu <= 0.4,
          ...
        
        1..2
        # failed 1 of 2 tests
    not ok 6 - Collect - process stat - cpu usage # time=417.617ms
 ```

---

Regarding native dependencies: does `@tensorflow/tfjs-node` ship with prebuilts? how much execution time are we going to lose by using JS-only?

**Comment by AndreasMadsen** [comment]
@mcollina @mafintosh I would love to see that. I tested the slow I/O example on the Pathfinders Weekly demo (using Zoom) and it worked fine. I didn't test the other examples, so I can't speak for that.

**Comment by AndreasMadsen** [comment]
All should be green now. I added a lot more tests to get 100% coverage. I also cleaned up the code such it is callback based.

**Comment by mafintosh** [comment]
Code looks fine, except I think the UX around the native optional dep is problematic. LGTM if @mcollina think that's okay though.

**Comment by AndreasMadsen** [comment]
> Code looks fine, except I think the UX around the native optional dep is problematic. 

That should be fixed by https://github.com/nearform/node-clinic-doctor/pull/149/files#diff-1a4f518b82da703b8bc1bcb843e8af39R12. Did you see the message in the latest version?

**Comment by mafintosh** [comment]
@AndreasMadsen ah, excellent!

**Comment by AndreasMadsen** [comment]
@kgryte It is not really clear what you are suggesting. – In any case, you can be assured that I didn't pull this model randomly out of a hat. Please check the R-script, if you want to see more details.

---

> Not clear to me that an HMM is the most computationally efficient approach for filtering out non-user application CPU spikes.

It literally takes 5 seconds. I think it is okay.

> For instance, single-pass algorithms exist which use a simple t-test (or G-test, if you know the variance) to detect state changes.

I'm well aware of both t-test and G-test. I have never heard of them being used as an alternative to HMM. Do you have a resource I can read?

> HMMs require significant state residency in order to be reliable. Meaning, you need large sample sizes. If state transition accuracy is desired, in a UI, users will not be able to choose an arbitrarily small window. Another way of saying this is that an HMM is not time-window invariant (an HMM will detect different state transitions in larger samples than in smaller samples); whereas, other algorithms have better properties in this regard.

I don't think it should be "time-window invariant". Such a model would not have a prior properbility and for sure there is a prior probability for which state the application is in. In this case it is highly unlikely that the application is doing garbage collection at the begining of the time-series.

> Not clear to me that CPU utilization abides by Gaussian emissions. Was normality tested? If so, other emissions distributions may be better suited; however, they may increase algorithm complexity.

CPU usage is typically gamma distributed. I think it could be done efficiently with gamma distributions, but I didn't spend the time deriving the Lagrange updates for that. From the research, a gaussian distribution appeard to be fine. Feel free to contribute with a gamma variant.

> HMMs are best used in time-series scenarios where the number of states is unknown.

For sure HMMs works best if the number of states are known. That is after all a hard model assumtion. I can't even imagine any model that would work best with an unkown number of states. How would that work? – At best, it would just be one more parameter to estimate.

Then, one can use log-likelihood (or if you want to get fancier and optimize over the hyperparameters, the "evidence") to determine the best fit (i.e., number of states).

I would think the likelihood will always increase with the number of states, as there is no penality for overparameterization. I did try a BIC for model selection, but got poor results.

> For cases where the number of states is known, not clear to me that this approach cannot be equivalently done using the relatively straightforward algorithm k-means clustering (which is used in the underlying implementation for HMM initialization), which, while subject to local minima, can be done in such a way to maximize the chance of reaching a global minima.

k-means is only used in the initialization of HMM, you could replace it with any clustering algorithm.  in my research, I did try a GMM as an alternative to HMM. GMM also works reasonably well, and yes it is simpler but there are cases where using a model that treats the data as a time-series is important.

> As is evident from the blog post, a naive HMM force fit with two states will not be able to distinguish between non-user application spikes (e.g., caused by GC, etc) and legitimate CPU spikes (e.g., caused by running a CPU intensive process, like fitting an HMM). 

This is valid. I have had the same thought. From our experimental results it doesn't appear to much of a concern, but it is something we will evaluate as we get more feedback. I doubt it will be a problem, as we are looking at the 90% quantile to see if is too small. If there are systematically many spikes, then likely there is no I/O issue.

> One way around this is to first build a model of non-user application behavior and then apply that model to unknown data and use a likelihood ratio test to determine whether the model can be better fit using an alternative model.

Would be nice. Unfortunetly we don't have the data for that.

> Alternatively, with added algorithmic complexity, one can design HMM algorithms which can distinguish between two states which differ in their rates alone (e.g., suppose GC spikes happened at rate r_1 and legitimate spikes happened with rate r_2, but both had the same state mean and variance). But again, not clear to me that such complexity is warranted, especially when simpler algorithms may suffice.

Unfortunetly the rate of GC is dependent on the application.

> Probably goes without saying and is mentioned in the blog post, the best approach would be to actually correlate events happening in the other threads which are affecting the CPU utilization profile, which means noting when, e.g., GC is triggered, and then aligning that event with CPU utilization data. The post mentions that this is difficult because one is sampled and the other is event driven. Not clear to me why the event data cannot be quantized (i.e., binned), thus forming a similarity sampled data set.

It is more than just sample vs events, there are simply V8 events missing from from the tracing data. Again, feel free to look at the R-script.

**Comment by kgryte** [comment]
@AndreasMadsen My general point is that I question whether using HMMs is a worthwhile and warranted approach. This stems from my experience using them for both discrete- and continuous-time time-series data where the application domains were biological time-series and, probably of more importance, server metrics, such as CPU utilization and other metrics which can be obtained from `/proc`.

Re: on-line change-point detection approaches. Yes, see older ion channel literature. Vassili is one such author.

Re: time-window invariance. You misunderstand the point. For what you are trying to do, you need only change-point detection, not state identification. HMMs perform a global analysis which falters for smaller samples. If states are rarely occupied, as would be the case as time-series length goes to `0`, then HMMs will struggle to meaningfully identify states. Yes, of course, you can force fit an HMM to have a specified number of states, but you can use many other algorithms to "find" what you want.

Re: number of states. Yes, you must use some metric which penalizes model complexity (e.g., AIC or BIC). But there are other approaches such as maximum evidence (see Bronson et al) which can be more robustly used for identifying the correct number of states. If BIC did not work for you, it is likely because a 2 state fit is unwarranted, given the data. This should be more obvious if, in fact, CPU utilization undergoes state transitions of varying duration, which I have actually seen in real server data. Meaning, a forced 2 state fit does not generalize.

Re: GC rates. Yes, of course. My point was that processes having rates `r_1` and `r_2` can be disambiguated based on rates alone, regardless as to whether these rates are known _a priori_. The point is that an HMM algorithm can be implemented which can account for states which do not differ in their emissions alone. The larger point, however, is that this is a rabbit hole. You can spend endless time designing more complex HMMs, but even a simple HMM seems like overkill for what is trying to be achieved.

Re: missing events. Sure. However, you should be able to significantly increase the reliability of even a naive approach by aligning events you do know and then running an analysis on the "stitched" time-series (i.e., a time-series where the aligned time segments are removed).

Re: performance. Yes, may take only `5` seconds now. However, the scaling properties of HMMs are not good, as I predict will be empirically discovered over time.

**Comment by AndreasMadsen** [comment]
> Re: on-line change-point detection approaches. Yes, see older ion channel literature. Vassili is one such author.

I assume this is such paper: http://rstb.royalsocietypublishing.org/content/352/1349/39.short - I will read that later. I'm a bit busy now.

> Re: missing events. Sure. However, you should be able to significantly increase the reliability of even a naive approach by aligning events you do know and then running an analysis on the "stitched" time-series (i.e., a time-series where the aligned time segments are removed).

Please look at the actual data. I've uploaded a HTML version of the Rmd script for you: https://github.com/nearform/node-clinic-doctor/blob/master/docs/cpu-model-dev/model.html

> Re: performance. Yes, may take only 5 seconds now. However, the scaling properties of HMMs are not good, as I predict will be empirically discovered over time.

Maybe you misunderstand. We fit a new model in every analysis. The length of the time series and the number of time-series doesn't scale. It is usually fixed in length and there is only one.
